### ISSUE TITLE
How to detect a <any> to <string> assignment?

### ISSUE THREAD

**Issue by Mika83AC** [Microsoft/TypeScript#16296]
Hello everyone,

after searching a while, I didn't found an answer for the following "issue":

```
let token: any = 123.456;
function test(input: string) {}
test(token);
```

How can I enforce the TS compiler to give a warning because of assigning <any> to a <string>?

I can't influence the definition of "token" (just a example here) because it's coming from another library. I just need to get warned that there is a possible conflict so that I can deal with this somehow in my code.  

Regards,
Michael

**Comment by TravCav** [comment]
Best i can tell, and someone correct me if I'm wrong, because of Union Types and the way they are handled, it checks to see if the value coming in has any of the types required. Since "any" essentially has all of them it doesn't complain.  My guess is we'd have to have some way to specify in the signature "only this type exactly". I think.

**Comment by ikatyang** [comment]
There is a tslint rule [`no-unsafe-any`](https://palantir.github.io/tslint/rules/no-unsafe-any/) ([test-sample](https://github.com/palantir/tslint/blob/master/test/rules/no-unsafe-any/test.ts.lint)) may what you want.

**Comment by trevorsg** [comment]
The `any` type can be assigned to and is assignable from all other types. It's pretty much a kill-switch for the type checker, so this by design. Instead of typing it as `any`, you should use a union type to cover all possible types. But you will still get a compiler error unless it can determine that the type is a `number` (or `any`).

    declare let token: string | number;
    function test(input: string) {}
    test(token);

Produces this error: 

Argument of type 'string | number' is not assignable to parameter of type 'string'.
  Type 'number' is not assignable to type 'string'.

So you could use:

    declare let token: string | number;
    function test(input: string) { }
    if (typeof token === "string") {
        test(token); // no error
    }

**Comment by Mika83AC** [comment]
Hey @trevorsg,

as described, I can't change the type definition of "token" as it's not my code. The code snipped was only an example

The real world problem is that I'm using nodejs +express + bodyparser and that it automatically sets the the `req.body.token` (as example) to the type `<any>`. So I've to handle what token is, sure, but as I'm assigning it to a function which can only take a `<string>` I was hoping to get a warning that there might be a problem and I've to handle it on my own.

But I guess this is as deadlock as bodyparser is doing the right thing (setting type to `<any>`, what else should he do), I'm doing the right thing setting my func parameter to `<string>` and TypeScript is doing the right thing by allowing `<any>` to be assigned to `<string>` because `<any>` is a override.

So there can't be a "solution" for this little trap, or?

Regards,
Michael

**Comment by TravCav** [comment]
Would it be reasonable to add something on near or around the type to specify "only this type" like 
`function test(input: [string]) {}`
or
`function test(input: $string) {}`

**Comment by NN---** [comment]
@trevorsg There are some functions that are typed as any in lib.d.ts such as Function.bind and there is nothing you can do with that now.
I started using TSLint and found a real bug with that.
If TS compiler had a switch to explicitly type assert 'any', the bug could be found much earlier.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
